### PR TITLE
ZCS-3705:SortBy read asc/desc does not work with mailbox

### PR DESCRIPTION
**Pull request by sneha-patil-synacor**
Added following sortBy support:
readAsc,readDesc

Updated soap doc
Added junit test

ant test-all passed

### CODE REVIEW

**Comment by fatkudu** [comment]
Added @ilyusha as a reviewer as I know he had some comments on the previous pull request.

File: store/src/java/com/zimbra/cs/db/DbSearch.java [hunk 1]
```diff
@@ -173,6 +175,10 @@ static String orderBy(SortBy sort, boolean alias) {
         if (sort.getKey() == SortBy.Key.NONE) { // no ORDER BY for NONE
             return "";
         }
+        if (sort.getKey() == SortBy.Key.UNREAD) {
+            return " ORDER BY " + (alias ? SORT_COLUMN_ALIAS : toSortField(sort))
+                + (sort.getDirection() == SortBy.Direction.DESC ? " DESC" : "") + ", mi.id DESC";
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/db/DbSearch.java]: Why mi.id DESC by default?  Interestingly, this is very close to a change I propose in https://github.com/Zimbra/zm-mailbox/pull/513.  Maybe I should change that to default to DESC or perhaps use the same direction as the first sort.  Would be interested in your thoughts.

Comment (sneha-patil-synacor) [reply store/src/java/com/zimbra/cs/db/DbSearch.java]: to keep chronological order(latest on top) always for group of read and unread mails.
So, first sort by unread flag(Asc/Desc) and sort by id(Desc)

Comment (fatkudu) [reply store/src/java/com/zimbra/cs/db/DbSearch.java]: Note that I've committed my changes, so you will need to rebase - sorry.

**Comment by ilyusha** [comment]
After looking more carefully at the search code, here is what I found:
In an index-first query operation, a large chunk of the index result set is consumed (potentially the whole thing). This is done by iteratively expanding the lucene result limit until no more matches are found (See `LuceneQueryOperation::getNextResultsChunk`). After each call to this method in `DBQueryOperation::luceneFirstGetNextChunk`, the newly found index results are used as filters on a new DB query, and the final DB results are appended to a running list.

What I'm worried about is that while the _actual_ order of the underlying index result set isn't used, as the DB query enforces its own sort, consuming it _in batches_ may run the risk of returning an incorrect final sort of the concatenated result list when the DB sort differs from the original index sort.

For a simple example, assume 10 index docs [1,2,...,10] that we want to return in lexicographic order. However, they are fetched from the index in batches of 4 in numeric order. That would give us four batches [1,2,3,4], [5,6,7,8], [9,10] passed to the DB query. Each batch is sorted lexicographically, resulting in the results being [1,2,3,4,5,6,7,8,9,10]. However, the true results should be [1,10,2,3,4,5,6,7,8,9].

It's entirely possible that I missed something in the search code or that I misunderstood a key step. If this really is a problem, however, I think it's outside of the scope for this ticket and should be addressed separately.

**Pull request closed without merging**